rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,431 @@
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_asset_amount, deserialize_fee
+from rotkehlchen.types import ApiKey, ApiSecret, ExchangeAuthCredentials, Location, Timestamp
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.misc import ts_now_in_ms, ts_sec_to_ms
+from rotkehlchen.utils.mixins.cacheable import cache_response_timewise
+from rotkehlchen.utils.mixins.lockable import protect_with_lock
+from rotkehlchen.utils.serialization import jsonloads_dict
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+API_KEY_ERROR_CODE_ACTION = {
+    '-1000': 'An unknown error occurred while processing the request.',
+    '-1001': 'The api key or secret is in wrong format.',
+    '-1002': 'The api key or secret is invalid.',
+}
+# Max limit for all API v2 endpoints
+API_MAX_LIMIT = 1000
+# user_transactions endpoint constants
+# Sort mode
+USER_TRANSACTION_SORTING_MODE = 'asc'
+# Starting `since_id`
+USER_TRANSACTION_MIN_SINCE_ID = 1
+# Trade type int
+USER_TRANSACTION_TRADE_TYPE = {2}
+# Asset movement type int: 0 - deposit, 1 - withdrawal
+USER_TRANSACTION_ASSET_MOVEMENT_TYPE = {0, 1}
+KNOWN_NON_ASSET_KEYS_FOR_MOVEMENTS = {
+    'datetime',
+    'id',
+    'type',
+    'fee',
+}
+
+
+class TradePairData(NamedTuple):
+    pair: str
+    base_asset_symbol: str
+    quote_asset_symbol: str
+    base_asset: AssetWithOracles
+    quote_asset: AssetWithOracles
+
+
+class Woo(ExchangeInterface):
+    """Woo exchange api docs:
+    https://docs.woo.org/#general-information
+
+    """
+    def __init__(
+            self,
+            name: str,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super().__init__(
+            name=name,
+            location=Location.WOO,
+            api_key=api_key,
+            secret=secret,
+            database=database,
+        )
+        self.base_uri = 'https://api.woo.org'
+        self.msg_aggregator = msg_aggregator
+        # NB: x-api-signature & x-api-timestamp change per request
+        # x-api-key is constant
+        self.session.headers.update({
+            'x-api-key': self.api_key,
+        })
+
+    def first_connection(self) -> None:
+        self.first_connection_made = True
+
+    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
+        changed = super().edit_exchange_credentials(credentials)
+        if credentials.api_key is not None:
+            self.session.headers.update({'x-api-key': credentials.api_key})
+        return changed
+
+    @protect_with_lock()
+    @cache_response_timewise()
+    def query_balances(self) -> ExchangeQueryBalances:
+        """Return the account balances on Woo """
```

Comment:
```suggestion
        """Return the account balances on Woo"""
```

also add a docstring with the possible errors that can be raised if any